Build: Prepare build for more script modules

sirreal opened the pull request:
## What?
Prepare webpack module build to handle more Script Module builds for use in WordPress.

- Rename the "interactivity" webpack build to "script modules".
- Output script-modules builds to `build-module` folder (adjacent to the `build` folder currently used for scripts).
- Add `wpScriptModulesExports` field and configure webpack to use it. This follows the same basic syntax as [package.json `exports` fields](https://nodejs.org/api/packages.html#exports). Multiple module entrypoints can be exposed per package. However, it remains a custom field, so it is clear that these entrypoints are not intended for general consumption. In the future, module-only packages (interactivity, interactivity-router) can switch to using exports directly and likely add `type: module`, but that's follow-up work 🙂 
- There are some difficulties with webpack recognizing `wpScriptModulesExports` directly, so packages are inspected programmatically in order to generated webpack script modules entrypoints.
- Adjust script module registration accordingly to find the generated script modules.

This is a first step towards building more script modules, such as `@wordpress/a11y` which may be used in https://github.com/WordPress/gutenberg/pull/62906.

https://github.com/WordPress/gutenberg/pull/65101 shows how this is used to build the `@wordpress/a11y` script module.

Related PRs:
- https://github.com/WordPress/gutenberg/pull/65101
- https://github.com/WordPress/gutenberg/pull/65123
- https://github.com/WordPress/wordpress-develop/pull/7304

## Why?

This prepares the build to produce more script modules.

## How?

See above.

## Testing Instructions

It's important to ensure that script modules are correctly registered. This should be covered by e2e tests, but testing, in particular of the affected blocks (image, query, etc.) on the frontend is appreciated.

sirreal committed:
Replace "interactivity" build with "script-modules"

This is a general script-modules build, not interactivity specific.

sirreal committed:
Prepare webpack script-modules to build wp-module field

gziolo commented on lib/interactivity-api.php:
```diff
@@ -16,14 +16,14 @@ function gutenberg_reregister_interactivity_script_modules() {
 	wp_register_script_module(
 		'@wordpress/interactivity',
-		gutenberg_url( '/build/interactivity/' . ( SCRIPT_DEBUG ? 'debug.min.js' : 'index.min.js' ) ),
+		gutenberg_url( '/build/script-modules/interactivity/' . ( SCRIPT_DEBUG ? 'debug.min.js' : 'index.min.js' ) ),
```
How about fully mirroring `build` folder inside top-level `build-module`?

```suggestion
		gutenberg_url( '/build-module/interactivity/' . ( SCRIPT_DEBUG ? 'debug.min.js' : 'index.min.js' ) ),
```

What cons do I miss?

sirreal replied:
I'm not sure, maybe there aren't problems 🤷 

I didn't think too much about this. I kept it in the same place but renamed from `interactivity` to `script-modules` but that's something I'm happy to explore.

gziolo replied:
This way we could keep two top-level
build folders mirroring also how it works for individual packages.

sirreal replied:
I've used that approach 👍

gziolo commented on tools/webpack/script-modules.js:
```diff
@@ -15,33 +15,37 @@ const { baseConfig, plugins } = require( './shared' );
+		'interactivity/debug': './packages/interactivity/src/debug',
+		'interactivity/router': './packages/interactivity-router',
+
+		'block-library/file': './packages/block-library/src/file/view.js',
```
Sharing my notes: `glob` for view files would help here with maintenance. Feel free to ignore them as I need to collect some ideas next to the code I see.

gziolo commented on tools/webpack/script-modules.js:
```diff
@@ -15,33 +15,37 @@ const { baseConfig, plugins } = require( './shared' );
 	},
 	resolve: {
 		extensions: [ '.js', '.ts', '.tsx' ],
+		mainFields: [ 'wp-module' ],
```
Can you elaborate on the idea here? I'm curious why that would be necessary?

gziolo replied:
Also, how about `wp-script-module` instead of `wp-module` to follow the nomenclature used in WP core?

gziolo commented on packages/interactivity/package.json:
```diff
@@ -24,6 +24,7 @@
 	},
 	"main": "build/index.js",
 	"module": "build-module/index.js",
+	"wp-module": "build-module/index.js",
```
I was thinking about how we could smoothly transition to ES Modules here. What if we start using [`"type": "module"`](https://nodejs.org/api/packages.html#type) and completely remove the obsolete support for CJS here. This way we could easily distinguish what the package build process should do about packages:

`"type": "module"` + `exports` - ES Module only
`"main": "some/path.js"` - CJS
`"module": "some/path.js"` - ES Module

More on [package entry points here](https://nodejs.org/api/packages.html#package-entry-points). I'm still processing it all, so this is more of me documenting the efforts.

gziolo replied:
Interesting fact, we use `module` field when deciding whether to build the package:

https://github.com/WordPress/gutenberg/blob/e40447e274390fa76ae01d2db19be2dc7db6504f/bin/packages/get-packages.js#L45

gziolo replied:
We never check whether CJS is required or not, but we only generate the folder in the production mode to speed up `npm run dev` as CJS isn't consumed by the Gutenberg plugin:

https://github.com/WordPress/gutenberg/blob/e40447e274390fa76ae01d2db19be2dc7db6504f/bin/packages/build-worker.js#L32-L37

gziolo commented:
I started looking at how to convert more packages to ESM, so I spent some time investigating how we could annotate such packages in a way where webpack could easily distinguish that from package.json files by using existing fields like `type`, `exports`, `module`, `main`. The bonus point would be we migrate to the modern way of publishing and consuming ESM. It might be completely unnecessary for what you are exploring, but I have been thinking about that for quite some time and it felt like a good place to share.

sirreal committed:
Move script-module assets to build-module directory

sirreal committed:
Make bundle-size workflow aware of build-module dir

sirreal commented:
@gziolo I had explored that in #60952 and used the `resolve.mainFields` configuration. I've set it to use `wp-module` (exclusively) so _modules_ will be required to add that field to their package.json to be picked up. I'm open to other options, but that seems to work for now.

https://github.com/WordPress/gutenberg/pull/65101 implements a `@wordpress/a11y` script module with the same API as the `wp-a11y` script.

sirreal committed:
Use wp-script-module main field

This better aligns with "Script Modules" naming in Core.

sirreal committed:
Try wp-script-module-exports approach

This custom "exports" setup allows multiple entrypoints while
maintaining a custom "WordPress Script Modules" setup. This is helpful
for packages like interactivity and block-library that have several
script modules to build.

sirreal committed:
Update script module registration accordingly

gziolo commented on packages/block-library/package.json:
```diff
@@ -30,6 +30,13 @@
 		"{src,build,build-module}/*/init.js"
 	],
+	"wp-script-module-exports": {
+		"./file/view": "./build-module/file/view.js",
```
It starts to look like `exports` from `package.json`. Nice! 👍🏻

sirreal replied:
Yes! I think this is going to be a very good direction. It moves in the direction of exports, but doesn't switch all the packages immediately.

It's also a custom field, and these are custom builds in general specific for use as WordPress Script Modules. It makes it clear that they're not intended for general consumption.

sirreal replied:
I also think we can move in the direction of generating the entrypoints in a standard way by reading package.json files.

sirreal replied:
Unfortunately, it doesn't work 🙃  (I thought it did but I'm not sure why that was).

Problems:
When entrypoints are paths (`./packages/block-library/file/view`), webpack never sees the package.json in order to find these paths.

>   ERROR in block-library/file/view
>  Module not found: Error: Can't resolve './packages/block-library/file/view' in '…'

When entrypoints are to modules (`@wordpress/block-library/file/view`) webpack resolves these correctly, but they're handled as externals. DEWP rejects these. We could fix DEWP, but even then we have entrypoints like `@wordpress/interactivity` which DEWP will externalize and there doesn't seem to be a good way to say "Externalize this _except_ when it's the entrypoint." Externalizing entrypoints is, needless to say, bad and broken 🙂 

So I don't think actually using this field as webpack `exportsFields` is going to work, but we can read this field and build a path to it programatically.

gziolo replied:
> So I don't think actually using this field as webpack exportsFields is going to work, but we can read this field and build a path to it programatically.

That sounds like a great first step with some thin compact layer in the webpack config that scans package.json files and produces entry points that are hardcoded today. For blocks you could eventually use pattern matching.

sirreal committed:
Script modules must build _after_ packages

sirreal committed:
Add interactivity-router to root dependencies

sirreal committed:
Try building entrypoints from package.json

sirreal committed:
Use the generated entrypoints

sirreal committed:
Update interactivity-router script module path

sirreal committed:
Fix script-module entry

sirreal commented:
I've added the "no core sync required" label. This should not impact Core right now, although these changes would be ported to Core in the future to support building modules correctly.

sirreal commented on package-lock.json:
```diff
@@ -46,6 +46,7 @@
 				"@wordpress/i18n": "file:packages/i18n",
 				"@wordpress/icons": "file:packages/icons",
 				"@wordpress/interactivity": "file:packages/interactivity",
+				"@wordpress/interactivity-router": "file:packages/interactivity-router",
```
This was omitted but should have been present.

gziolo commented:
This is looking awesome. 👏 

It needs another rebase. I will took it for a spin tomorrow and take a final look at the code. I was happy to see that the project is configured to ignore all `build-module` folders. We also need to audit some common scripts to ensure they clean the newly added folder, some example that comes to my mind:
https://github.com/WordPress/gutenberg/blob/07c7821b07e5d2b2abaa86f6737dd486da26383e/package.json#L281

Thid one is not impacted and I wasn't able to locate anything problematic through a quick scan of this file.

gziolo commented on tools/webpack/script-modules.js:
```diff
@@ -0,0 +1,142 @@
+const WORDPRESS_NAMESPACE = '@wordpress/';
+const { createRequire } = require( 'node:module' );
+
+const rootPath = `${ __dirname }/../../`;
```
This can fail on windows machines. @t-hamano, are you able to run smoke tests?

sirreal replied:
Good catch, I'll use the appropriate methods to construct paths.

sirreal replied:
I've updated to use URLs where appropriate, it seems nicer but I would appreciate if someone on Windows gives it a test 🙏

gziolo commented on webpack.config.js:
```diff
@@ -3,14 +3,14 @@
 	...blocksConfig,
-	interactivity,
 	packagesConfig,
+	scriptModules, // Script modules build depends on the package build.
```
Interesting, what is the reason here? I’m curious.

sirreal replied:
I think this is wrong and I'll revert it. I was confusing the build of packages with this "packagesConfig" as I debugged some of the entrypoint issues.

sirreal committed:
Revert webpack configs reordering

sirreal committed:
Use URLs for improved cross platform support

sirreal commented:
Rebased to fix conflicts, ready for review.

gziolo commented:
Some places to audit whether we should extend support for `build-module`:

This is must have:
https://github.com/WordPress/gutenberg/blob/2f939a7f9047ed9b632ac4fb0d593d0cf9588ce9/bin/build-plugin-zip.sh#L81-L90

I'm not sure what this is about:
https://github.com/WordPress/gutenberg/blob/2f939a7f9047ed9b632ac4fb0d593d0cf9588ce9/.github/workflows/unit-test.yml#L120-L138

--- 

Potential improvement for later - definitely a separate PR which I can open at some point. The following function doesn't have to be about Interactivity API anymore:

https://github.com/WordPress/gutenberg/blob/2f939a7f9047ed9b632ac4fb0d593d0cf9588ce9/lib/interactivity-api.php#L12-L30

Instead of hardcoding everything, we could create a single manifest file from webpack containing all script modules and register them in the loop as we do in WP core:

https://github.com/WordPress/wordpress-develop/blob/afeaf7e8ce7ad7e7030f1b95ef7d7515810a6429/src/wp-includes/script-loader.php#L272-L283

gziolo commented:
I see that the `view.js` files for blocks are produced twice:

<img width="1867" alt="Screenshot 2024-09-11 at 10 32 42" src="https://github.com/user-attachments/assets/193a5df2-1e6c-4433-ad37-c2fc98b30762">

It's a preexisting issue as that was missed during the refactoring. It's probably some artifacts from WP 6.4. It would also explain why you run into issues with the order of configs.

Code to refactor is here:

https://github.com/WordPress/gutenberg/blob/f5e2162648276d8cf37b7597636cbd189a740283/tools/webpack/blocks.js#L53-L80

It probably means there would be no entry points, so we will also have to think how to structure configs. Having blocks independently from entry points was the strategy in the old world of scripts.

sirreal commented on tools/webpack/script-modules.js:
```diff
@@ -0,0 +1,130 @@
+const WORDPRESS_NAMESPACE = '@wordpress/';
+const { createRequire } = require( 'node:module' );
+
+const rootPath = new URL( '..', `file://${ __dirname }` );
```
This is now a URL:

```suggestion
const rootURL = new URL( '..', `file://${ __dirname }` );
```

gziolo commented:
I tried the following to remove obsolete view files in the `build/block-library/blocks/` folder and it seems to work fine:

```diff
diff --git a/tools/webpack/blocks.js b/tools/webpack/blocks.js
index fa0c6123db..b496164e39 100644
--- a/tools/webpack/blocks.js
+++ b/tools/webpack/blocks.js
@@ -3,7 +3,6 @@
  */
 const CopyWebpackPlugin = require( 'copy-webpack-plugin' );
 const { join, sep } = require( 'path' );
-const fastGlob = require( 'fast-glob' );
 const { realpathSync } = require( 'fs' );
 
 /**
@@ -17,13 +16,6 @@ const { PhpFilePathsPlugin } = require( '@wordpress/scripts/utils' );
  */
 const { baseConfig, plugins, stylesTransform } = require( './shared' );
 
-/*
- * Matches a block's filepaths in the form build-module/<filename>.js
- */
-const blockViewRegex = new RegExp(
-	/build-module\/(?<filename>.*\/view.*).js$/
-);
-
 /**
  * We need to automatically rename some functions when they are called inside block files,
  * but have been declared elsewhere. This way we can call Gutenberg override functions, but
@@ -50,40 +42,11 @@ function escapeRegExp( string ) {
 	return string.replace( /[\\^$.*+?()[\]{}|]/g, '\\$&' );
 }
 
-const createEntrypoints = () => {
-	/*
-	 * Returns an array of paths to block view files within the `@wordpress/block-library` package.
-	 * These paths can be matched by the regex `blockViewRegex` in order to extract
-	 * the block's filename.
-	 *
-	 * Returns an empty array if no files were found.
-	 */
-	const blockViewScriptPaths = fastGlob.sync(
-		'./packages/block-library/build-module/**/view*.js'
-	);
-
-	/*
-	 * Go through the paths found above, in order to define webpack entry points for
-	 * each block's view.js file.
-	 */
-	return blockViewScriptPaths.reduce( ( entries, scriptPath ) => {
-		const result = scriptPath.match( blockViewRegex );
-		if ( ! result?.groups?.filename ) {
-			return entries;
-		}
-
-		return {
-			...entries,
-			[ result.groups.filename ]: scriptPath,
-		};
-	}, {} );
-};
-
 module.exports = [
 	{
 		...baseConfig,
 		name: 'blocks',
-		entry: createEntrypoints(),
+		entry: {},
 		output: {
 			devtoolNamespace: 'wp',
 			filename: './build/block-library/blocks/[name].min.js',
```

Happy to open a separate PR.

sirreal commented:
> I tried the following to remove obsolete view files in the build/block-library/blocks/ folder and it seems to work fine

I would like to keep that work separate as it's preexisting and on its own it will be easier to work with.

I will look at build-plugin-zip, the list of `build/` files seems to be everything so maybe the `build` and `build-module` files should just be included in the zip entirely.

sirreal committed:
Include build-module in plugin zip

sirreal committed:
URL renames and cleanup

gziolo commented on bin/build-plugin-zip.sh:
```diff
@@ -98,6 +98,7 @@ zip -r gutenberg.zip \
 	packages/block-serialization-default-parser/*.php \
 	post-content.php \
 	$build_files \
+	build-module \
```
I don't remember why there was pattern matching used for files in the `build` folder. @youknowriad, is there anything important we might miss when including the entire `build-module`? At the moment, we want all files to be included from there.

sirreal replied:
I created related PR https://github.com/WordPress/gutenberg/pull/65064

gziolo reviewed (approved):
Code looks great, I spent some time testing both `npm run dev` and `npm run build`. Everything is working as before. I confirmed that the key Interactivity API features: the lightbox in the Image block and instant pagination in the Query block work correctly. I also tested with the [Playground instance](https://playground.wordpress.net/#%7B%22landingPage%22:%22/wp-admin/%22,%22features%22:%7B%22networking%22:true%7D,%22steps%22:%5B%7B%22step%22:%22login%22,%22username%22:%22admin%22,%22password%22:%22password%22%7D,%7B%22step%22:%22mkdir%22,%22path%22:%22/wordpress/pr%22%7D,%7B%22step%22:%22writeFile%22,%22path%22:%22/wordpress/pr/pr.zip%22,%22data%22:%7B%22resource%22:%22url%22,%22url%22:%22/plugin-proxy.php?org=WordPress&repo=gutenberg&workflow=Build%2520Gutenberg%2520Plugin%2520Zip&artifact=gutenberg-plugin&pr=65064%22,%22caption%22:%22Downloading%20Gutenberg%20PR%2065064%22%7D,%22progress%22:%7B%22weight%22:2,%22caption%22:%22Applying%20Gutenberg%20PR%2065064%22%7D%7D,%7B%22step%22:%22unzip%22,%22zipPath%22:%22/wordpress/pr/pr.zip%22,%22extractToPath%22:%22/wordpress/pr%22%7D,%7B%22step%22:%22installPlugin%22,%22pluginZipFile%22:%7B%22resource%22:%22vfs%22,%22path%22:%22/wordpress/pr/gutenberg.zip%22%7D%7D%5D%7D) that confirms that changes to the zip file provided for the plugin is fully functional.

There are some open questions to resolve, and we need to confirm that everything works correctly on Windows before proceeding. I'm approving, as the rest is not under my control.

t-hamano commented:
I checked out this branch on my Windows host OS and ran `npm run build`, and the `build-module` directory seems to be generated correctly:

![build-module](https://github.com/user-attachments/assets/977dcc84-6105-4f30-9f1e-f2335d119fc6)

However, I noticed that the PHP files for the blocks are not generated in the `build/block-library/blocks/` directory:

| Build on Windows host OS | Build on WSL2 (Ubuntu) |
|--------|--------|
| ![windows](https://github.com/user-attachments/assets/4028cc9b-83c7-4314-94ae-bff5d74b3a9f) |  ![ubuntu](https://github.com/user-attachments/assets/31fc4da9-041d-4cdf-bcbc-ad0d0bddda87) | 

Because I cannot build the PHP files for the blocks changed in this PR, I cannot test the behavior of the Image block and other blocks on the front end.

However, this problem also occurs in the trunk. I have not yet been able to investigate this in detail, but it might be related to #63098.

In any case, the `build-module` directory is building correctly on the Windows host OS, so is it okay to ship this PR? I'll look into the issue of PHP files not being generated in more detail later.

sirreal commented:
Here's the expected output of listing `build-module` after `npm run build`:

```
build-module:
block-library
interactivity
interactivity-router

build-module/block-library:
file
image
navigation
query
search

build-module/block-library/file:
view.js
view.min.asset.php
view.min.js
view.min.js.map

build-module/block-library/image:
view.js
view.min.asset.php
view.min.js
view.min.js.map

build-module/block-library/navigation:
view.js
view.min.asset.php
view.min.js
view.min.js.map

build-module/block-library/query:
view.js
view.min.asset.php
view.min.js
view.min.js.map

build-module/block-library/search:
view.js
view.min.asset.php
view.min.js
view.min.js.map

build-module/interactivity:
debug.js
debug.min.asset.php
debug.min.js
debug.min.js.map
index.js
index.min.asset.php
index.min.js
index.min.js.map

build-module/interactivity-router:
index.js
index.min.asset.php
index.min.js
index.min.js.map
```

sirreal commented:
Thanks @t-hamano! It looks like this is OK to move ahead if it's building on Windows 👏

gziolo commented:
Opened a follow-up PR to clean up the unused legacy scripts for core blocks:

- https://github.com/WordPress/gutenberg/pull/65268
